### PR TITLE
Downloads files from GCS 

### PR DESCRIPTION
**Pull request by shelbychen97**
This change is [Reviewable](https://reviewable.io/reviews/m-lab/annotation-service/9)

### CODE REVIEW

**Comment by yachang** [comment]
Review status: 0 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[README.md, line 2 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcTQ79BtoUUN_m3DWN:-KqcTQ79BtoUUN_m3DWO:b-v6yx04) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/README.md#L2)):*
> ```Markdown
> # annotation-service
> Annotation integration service for M-Lab datai.
> ```

typo, get rid of "i"

---

*[README.md, line 4 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcTWQB_2ElmD4kkgpL:-KqcTWQB_2ElmD4kkgpM:b9pbzo3) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/README.md#L4)):*
> ```Markdown
> Annotation integration service for M-Lab datai.
> 
> Downloader package reads in a specified CSV IPv4 database from GCS. 
> ```

Can your code read IPv6 package now? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
Review status: 0 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[downloader/downloadGCS.go, line 8 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcTxuFnaJn60zMiLqU:-KqcTxuFnaJn60zMiLqV:bbzsfom) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/downloader/downloadGCS.go#L8)):*
> ```Go
> 	"golang.org/x/net/context"
> 	"cloud.google.com/go/storage"
> 	"github.com/m-lab/annotation-service/parser"
> ```

space between "github.com..." and above lines.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
Review status: 0 of 10 files reviewed at latest revision, 4 unresolved discussions.

---

*[downloader/downloadGCS.go, line 14 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcUV0RsNBVFUUDHCzV:-KqcUV0RsNBVFUUDHCzW:bq86ytb) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/downloader/downloadGCS.go#L14)):*
> ```Go
> 
> //Creates list of IP address Nodes 
> func InitializeTable(ctx context.Context, GCSFolder,GCSFile string) *[]parser.Node{
> ```

return (*[]parser.Node, error), and the calling function can check whether error is nil.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[downloader/downloadGCS.go, line 43 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcVEZMEu9isqp4r4KJ:-KqcVEZMEu9isqp4r4KK:be88cxp) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/downloader/downloadGCS.go#L43)):*
> ```Go
> 	bkt := client.Bucket(bucket)
> 
> 	obj := bkt.Object(bucketObj)
> ```

above 2 lines can be one line:

obj := client.Bucket(bucket).Object(bucketObj)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
Review status: 0 of 10 files reviewed at latest revision, 6 unresolved discussions.

---

*[downloader/downloadGCS_test.go, line 20 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcYU3VVfUG2VSr9uAc:-KqcYU3VVfUG2VSr9uAd:bewquth) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/downloader/downloadGCS_test.go#L20)):*
> ```Go
> 		}
> 		defer done()
> 		geoData := downloader.InitializeTable(ctx,"test-annotator-sandbox","annotator-data/GeoIPCountryWhoisSAMPLE.csv") 
> ```

run "go fmt"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by shelbychen97** [comment]
Review status: 0 of 12 files reviewed at latest revision, 6 unresolved discussions.

---

*[README.md, line 2 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcTQ79BtoUUN_m3DWN:-KqdlcRa87iDSO-8PiXy:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/README.md#L2)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

typo, get rid of "i"
</blockquote></details>

Done.

---

*[README.md, line 4 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcTWQB_2ElmD4kkgpL:-Kqdlcnq6FJvNhvRuB6Z:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/README.md#L4)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

Can your code read IPv6 package now? 
</blockquote></details>

Done.

---

*[downloader/downloadGCS.go, line 8 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcTxuFnaJn60zMiLqU:-Kqdldgmo7kw5MbQh9PJ:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/downloader/downloadGCS.go#L8)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

space between "github.com..." and above lines.
</blockquote></details>

Done.

---

*[downloader/downloadGCS.go, line 14 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcUV0RsNBVFUUDHCzV:-Kqdle9WkGgAT54GnBT6:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/downloader/downloadGCS.go#L14)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

return (*[]parser.Node, error), and the calling function can check whether error is nil.
</blockquote></details>

Done.

---

*[downloader/downloadGCS.go, line 43 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcVEZMEu9isqp4r4KJ:-Kqdlej-mVWKZb6eJC5Y:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/downloader/downloadGCS.go#L43)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

above 2 lines can be one line:

obj := client.Bucket(bucket).Object(bucketObj)
</blockquote></details>

Done.

---

*[downloader/downloadGCS_test.go, line 20 at r1](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqcYU3VVfUG2VSr9uAc:-Kqdm2D1tGZdAXRVkset:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/c8c775ce1d04bf41810b4c982013768f6d766138/downloader/downloadGCS_test.go#L20)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

run "go fmt"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

File: parser/geo-ip.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
+package parser
```
Comment (yachang) [inline parser/geo-ip.go]: Add documentation to summarize the main functionality of this file.

File: annotator.sh [hunk 2]
```diff
@@ -5,5 +5,5 @@
 $TRAVIS_BUILD_DIR/travis/install_gcloud.sh app-engine-python app-engine-go
 echo $sandbox_service_key | base64 --decode -i > ${HOME}/gcloud-service-key.json
 gcloud auth activate-service-account --key-file ${HOME}/gcloud-service-key.json
-export GOOGLE_APPLICATION_CREDENTIALS="${HOME}/gcloud-service-key.json" && export APPENGINE_DEV_APPSERVER="${HOME}/google-cloud-sdk/bin/dev_appserver.py" && source "${HOME}/google-cloud-sdk/path.bash.inc" && go test -v ./...
+export GOOGLE_APPLICATION_CREDENTIALS="${HOME}/gcloud-service-key.json" && export APPENGINE_DEV_APPSERVER="${HOME}/google-cloud-sdk/bin/dev_appserver.py" && source "${HOME}/google-cloud-sdk/path.bash.inc"
```
Comment (yachang) [inline annotator.sh]: If you set GOOGLE_APPLICATION_CREDENTIALS etc here, why do you need to redo them in .travis.yaml?

File: downloader/downloadGCS_test.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+import (
+	"google.golang.org/appengine/aetest"
+	"os"
+	"testing"
+
+	"github.com/m-lab/annotation-service/downloader"
+	"github.com/m-lab/annotation-service/parser"
+)
+
+func TestInitilizationTableIPv4(t *testing.T) {
```
Comment (yachang) [inline downloader/downloadGCS_test.go]: This function and the below function TestInitilizationTableIPv6() has too much common code.

Can you try to refactor and reduce the dup code?

File: parser/geo-ip.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+func CreateListIPv4(reader io.Reader) ([]Node, error) {
+	list := []Node{}
+	r := csv.NewReader(reader)
+	r.TrimLeadingSpace = true
+	for {
+		record, err := r.Read()
+		if err == io.EOF {
+			break
+		}
+		var newNode Node
```
Comment (yachang) [inline parser/geo-ip.go]: Check the len of record.

Add unittest for corrupted input data.

File: parser/geo-ip.go (hunk 1 above)
Comment (shelbychen97) [inline parser/geo-ip.go]: Done

File: annotator.sh (hunk 2 above)
Comment (shelbychen97) [inline annotator.sh]: Done

File: downloader/downloadGCS_test.go (hunk 3 above)
Comment (shelbychen97) [inline downloader/downloadGCS_test.go]: Done

File: parser/geo-ip.go (hunk 4 above)
Comment (shelbychen97) [inline parser/geo-ip.go]: Done

**Comment by yachang** [comment]
Review status: 0 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[downloader/downloadGCS_test.go, line 53 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxJ5IPK6obuDecZuEj:-KqxJ5IPK6obuDecZuEk:bwr6vq0) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/downloader/downloadGCS_test.go#L53)):*
> ```Go
> 	}
> 	r, _ := os.Open("testdata/GeoLiteCityv6SAMPLE.csv")
> 	LocalGeoData, _ := parser.CreateList(r,6)
> ```

Check whether there is error returned here, and call t.Errorf() if there is one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
Review status: 0 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[parser/geo-ip.go, line 1 at r2](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqZOYnwWVniqnwO2nES-r2-1:-KqxJGgsiXRKwXLlQ3yA:bc0b2s1) ([raw file](https://github.com/m-lab/annotation-service/blob/61856cf0170bbd677f7563ca10b2d8ef0ec87ae2/parser/geo-ip.go#L1)):*
<details><summary><i>Previously, shelbychen97 (Shelby Chen ) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Put the documentation at the first line, like:

// Depending on whether user input was an IPv4 or IPv6 IPaddress,
// respective database file will be read in and a list of Nodes will be created
// Each node contains a geo-location and its range of IP addresses.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
Review status: 0 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[parser/geo-ip_test.go, line 16 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxKaTFPwHN2QtU28op:-KqxKaTGpOPQQxTgUGCg:brj8esy) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/parser/geo-ip_test.go#L16)):*
> ```Go
> 	list, err := parser.CreateList(r,4)
> 	if err != nil {
> 		t.Errorf("Error in creating list") 
> ```

The ending dot can compile?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
Review status: 0 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[downloader/downloadGCS_test.go, line 22 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxLYD-twsbVv_b_TtP:-KqxLYD-twsbVv_b_TtQ:b-vgfro8) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/downloader/downloadGCS_test.go#L22)):*
> ```Go
> 		t.Errorf("InitializeTable failed.")
> 	}
> 	r, _ := os.Open("testdata/GeoIPCountryWhoisSAMPLE.csv")
> ```

check err here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
Review status: 0 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[downloader/downloadGCS_test.go, line 24 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxMNHeNFrYD5E6qtC8:-KqxMNHeNFrYD5E6qtC9:bmh961c) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/downloader/downloadGCS_test.go#L24)):*
> ```Go
> 	r, _ := os.Open("testdata/GeoIPCountryWhoisSAMPLE.csv")
> 	LocalGeoData, _ := parser.CreateList(r,4)
> 	for index, element := range geoData {
> ```

The following codes are dup with TestInitilizationTableIPv6(). Create a new func and remove the dup code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[parser/geo-ip_test.go, line 44 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxNiiOXw3iLE7hi_Qg:-KqxNiiOXw3iLE7hi_Qh:b-xk4o81) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/parser/geo-ip_test.go#L44)):*
> ```Go
> 		},
> 	}
> 	for index, element := range list {
> ```

This for loop are dup code. Try to refactor and remove dup code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by shelbychen97** [comment]
Review status: 0 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[downloader/downloadGCS_test.go, line 22 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxLYD-twsbVv_b_TtP:-Kqxu-8hnEdef0bWuO4e:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/downloader/downloadGCS_test.go#L22)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

check err here
</blockquote></details>

Done.

---

*[downloader/downloadGCS_test.go, line 24 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxMNHeNFrYD5E6qtC8:-Kqxu01D_IM8S0e70SPR:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/downloader/downloadGCS_test.go#L24)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

The following codes are dup with TestInitilizationTableIPv6(). Create a new func and remove the dup code.
</blockquote></details>

Done.

---

*[downloader/downloadGCS_test.go, line 53 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxJ5IPK6obuDecZuEj:-KqxxVKuBgOhH1vtJbb8:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/downloader/downloadGCS_test.go#L53)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

Check whether there is error returned here, and call t.Errorf() if there is one.
</blockquote></details>

Done.

---

*[parser/geo-ip.go, line 1 at r2](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqZOYnwWVniqnwO2nES-r2-1:-KqxvdeLW1mBHeD7X3Oy:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/61856cf0170bbd677f7563ca10b2d8ef0ec87ae2/parser/geo-ip.go#L1)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

Put the documentation at the first line, like:

// Depending on whether user input was an IPv4 or IPv6 IPaddress,
// respective database file will be read in and a list of Nodes will be created
// Each node contains a geo-location and its range of IP addresses.
</blockquote></details>

Done.

---

*[parser/geo-ip_test.go, line 16 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxKaTFPwHN2QtU28op:-Kqxvkv2GzWBbigQHQ0n:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/parser/geo-ip_test.go#L16)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

The ending dot can compile?
</blockquote></details>

Done.

---

*[parser/geo-ip_test.go, line 44 at r3](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-KqxNiiOXw3iLE7hi_Qg:-Kqxvn5pUSF67VFo8ubU:b-896fix) ([raw file](https://github.com/m-lab/annotation-service/blob/de84fd95b2e43947fab40b275ee4502647238b3f/parser/geo-ip_test.go#L44)):*
<details><summary><i>Previously, yachang wrote…</i></summary><blockquote>

This for loop are dup code. Try to refactor and remove dup code.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9)*

**Comment by yachang** [comment]
:lgtm:

---

Review status: 0 of 13 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/m-lab/annotation-service/9#-:-KqyB7g0vHgVfebExCVt:bnfp4nl)*

**Pull request merged**